app/models/puppetclass.rb

```diff
@@ -110,8 +107,8 @@ def all_host_ids
     ids.uniq
   end
 
-  def update_total_hosts
-    update_attribute(:total_hosts, all_host_ids.size)
+  def hosts_count
+    all_host_ids.count
```

Comment:
Hmm, the code that builds `all_host_ids` is quite complex, but it might be possible to optimize it a bit - at least do two queries and sum them up.